### PR TITLE
[Bug fix] fix(eltwise): overflow-safe logaddexp/logaddexp2 via native SFPU kernel

### LINKED ISSUE #52037: [Bounty $1500] logaddexp / logaddexp2: overflow-safe reformulation
## Summary

`ttnn.logaddexp` and `ttnn.logaddexp2` are composed as `log(exp(a) + exp(b))` literally, so they
return `±inf` for arguments whose exact result is an ordinary finite number. The whole point of a
log-sum-exp primitive is that it does not overflow; this one overflows at `|a| > 88.7`, which is
inside the normal working range of any log-domain computation.

Both directions fail: `logaddexp(100, 0)` returns `inf` where the exact answer is `100`, and
`logaddexp(-100, -100)` returns `-inf` where the exact answer is `-99.31`.

There is no representable-range argument available here, because the result is boxed in by its
own inputs:

```
max(a, b)  <=  logaddexp(a, b)  <=  max(a, b) + ln 2
```

If both arguments are finite the answer is finite and within 0.694 of one of them, so an `inf`
return is never a range limitation — it is always the intermediate.

## Current state measured on ttsim

ttsim 1.9.7, `Arch.BLACKHOLE`, `ttnn.logaddexp` called directly:

| a | b | chip | exact | |
|---|---|---|---|---|
| 100 | 0 | `inf` | 100.000000 | ✗ |
| 89 | 0 | `inf` | 89.000000 | ✗ |
| 90 | 89 | `inf` | 90.313263 | ✗ |
| 100 | 100 | `inf` | 100.693146 | ✗ |
| 200 | 199 | `inf` | 200.313263 | ✗ |
| −100 | −100 | `−inf` | −99.306854 | ✗ |
| 88 | 88 | 88.693146 | 88.693146 | ok |
| 5 | 3 | 5.126928 | 5.126928 | ok |

The threshold is sharp, and it is the `exp` overflow point rather than anything about the pair:

```
a =  88.70, b = 0   ->  88.699997   exact  88.699997    ok
a =  89.00, b = 0   ->  inf         exact  89.000000    fails
```

`bfloat16` behaves identically — `log(FLT_MAX)` is 88.72 and `log(BF16_MAX)` is 88.72, so the
narrower dtype changes nothing.

`logaddexp2` fails the same way at its own threshold, since it is built on `exp2` rather than
`exp`. Last good and first failing argument for `a = b`, both dtypes:

| op | positive | negative |
|---|---|---|
| `logaddexp` | 88 ok, 90 `inf` | −86 ok, −88 `−inf` |
| `logaddexp2` | 126 ok, 128 `inf` | −126 ok, −128 `−inf` |

`logaddexp2(1e5, 1e5)` returns `inf`; the exact answer is `100001`, which is an ordinary fp32
number.

## Root cause

The composition is a fixed `unary-pre / binary / unary-post` pipeline, and it names the two
overflowing steps explicitly:

https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_utils.cpp#L250-L263

```cpp
// log( exp(a) + exp(b) )
case BinaryOpType::LOGADDEXP:
    process_lhs  = unary::UnaryOpType::EXP;
    process_rhs  = unary::UnaryOpType::EXP;
    binary_op    = EnumT::ADD;
    postprocess  = unary::UnaryOpType::LOG;
```

`exp(a)` is evaluated before anything else can bound it, so it saturates and `log(inf)` is `inf`.
The underflow case is the mirror: both `exp` terms flush to zero and `log(0)` is `−inf`.

The same shape is repeated in two more places, so a fix has to cover all three:

- `ttnn/cpp/ttnn/operations/eltwise/binary/common/binary_op_utils.cpp#L130-L138` — FPU path,
  `EXP` as `PRE_IN0_0` / `PRE_IN1_0` around `add_tiles`, `LOG` as postprocess.
- `ttnn/cpp/ttnn/operations/eltwise/binary/common/binary_op_utils.cpp#L401-L410` — SFPU path,
  same defines around `add_binary_tile`.
- `LOGADDEXP2` is the identical pattern with `EXP2` / `LOG2` in each of the three.

This is host-side composition rather than kernel code, so it is not architecture-specific:
Wormhole, Blackhole and the Quasar copy all build the same pipeline.

## Direction

The standard identity keeps everything in range:

```
logaddexp(a, b) = max(a, b) + log1p(exp(-|a - b|))
```

`|a − b|` is non-negative, so `exp` of its negation is in `(0, 1]` and cannot overflow; the
result inherits its magnitude from `max(a, b)` rather than from `exp`. The primitives are already
in the tree — `ttnn.log1p` and `ttnn.maximum` both exist — but this does not fit the current
`pre/binary/post` slot shape, so whether it lands as a dedicated fused SFPU binary tile op
(alongside `lcm_tile`, `mul_binary_tile`) or as a restructured composition is a design and
performance call I am deliberately leaving to you.

For reference, the identity evaluated in fp32 gives the exact answer on every case above,
including `logaddexp(1000, 999) = 1000.313232`.

## Success criteria

1. `logaddexp(a, b)` and `logaddexp2(a, b)` are finite whenever the exact result is finite, for
   `|a|, |b| <= 1e4`, in both `float32` and `bfloat16`.
2. Relative error against `torch.logaddexp` stays within the op's existing tolerance over that
   range — the identity above is exact to fp32 rounding, so this should not cost accuracy.
3. Values in the currently-working range (`|a|, |b| < 88.7`) are unchanged within tolerance.
4. All three composition sites and the `LOGADDEXP2` variants are covered, with a sweep entry that
   samples beyond `±88.7` — the existing coverage never does, which is why this has stayed open.

## Why this matters

Log-sum-exp exists precisely to be the overflow-safe way to add in the log domain. A version that
overflows at 88.7 cannot be used for the thing it is for: numerically-stable softmax denominators,
log-likelihood accumulation, mixture-of-experts routing. The failure is silent — `inf` propagates
through the following ops and surfaces as `NaN` somewhere much later.

Worth separating from #26993, which is closed and lists `logaddexp` under the `exp` re-evaluation:
that work was about ULP accuracy of `exp` itself and did not change the composition. No amount of
precision in `exp` prevents `exp(100)` from overflowing.

### PR DESCRIPTION
**Pull request by thwaitsusa0316**
### PR Category
Bug fix

### Summary
`ttnn.logaddexp` and `ttnn.logaddexp2` silently return `inf` for large inputs
because the naive decomposition `log(exp(a) + exp(b))` saturates at |a| or |b| > ~88.7
(the FP32 overflow threshold of `exp`).

This PR fixes both ops using the numerically stable identity:
  logaddexp(a,b)  = max(a,b) + log(1 + exp(-|a-b|))
  logaddexp2(a,b) = max(a,b) + log2(1 + 2^{-|a-b|})

Because exp(-|a-b|) has argument ≤ 0, its result is in (0,1] and can never
overflow regardless of the magnitude of a or b. The log argument is in (1,2],
so log is also well-behaved.

Implementation:
- Adds BinaryOp::LOGADDEXP/LOGADDEXP2 to ckernel_defs.h (wormhole_b0 + blackhole)
- Implements the stable kernel cases in ckernel_sfpu_binary.h (both architectures)
- Adds logaddexp.h / logaddexp2.h compute API headers (mirrors xlogy.h pattern)
- Wires SfpuBinaryOp::LOGADDEXP/LOGADDEXP2 through binary_ng_utils and binary_op_utils
- Removes the broken unary-compose EXP+ADD+LOG pipeline from both code paths

Fixes #52037

### Notes for reviewers
- The two kernel cases in ckernel_sfpu_binary.h are the core change — the rest is wiring.
- sfpi::abs() is used for |a-b|; v_if (in1 > in0) for the max.
- _calculate_exponential_body_ and _calculate_log_body_ are reused from the existing XLOGY case.
- log2(x) = log(x) * 1.4426950f since _calculate_log2_body_ does not exist.
- The legacy FPU path in get_defines() retains the old decomposition; only the SFPU and binary_ng paths are fixed here.
- Hardware validation requires Wormhole n300s; CI results will confirm correctness.

### CODE REVIEW

**Comment by EazyHood** [comment]
Nice — the identity and the routing both look right to me, and covering `logaddexp2` in the same
pass is more than I had.

One measurement on the kernel, in case it is useful before this lands. The exponential you picked
costs about five orders of magnitude of accuracy, and swapping it is a one-line change.

`_calculate_exponential_body_` returns `255/256` rather than `1` at zero, and since the argument
here is `-|a-b|`, that error lands directly on `log(1 + exp_neg)` and then on the result. Measured
on ttsim 1.9.7 (`Arch.BLACKHOLE`, `float32`), your kernel as written:

| a | b | this PR | exact | rel err |
|---|---|---|---|---|
| 0 | 0 | 0.676983 | 0.693147 | **2.33e−02** |
| 5 | 3 | 5.128298 | 5.126928 | 2.67e−04 |
| 100 | 100 | 100.676987 | 100.693147 | 1.60e−04 |
| −100 | −100 | −99.323013 | −99.306853 | 1.63e−04 |
| 100 | 0 | 100.000999 | 100.000000 | 9.99e−06 |
| 1000 | 999 | 1000.313293 | 1000.313262 | 3.18e−08 |

The error is largest when `a ≈ b`, which is where `exp(-|a-b|)` is closest to 1 and the
approximation is furthest off. It shrinks as the arguments separate, because `exp_neg` goes to
zero and the result collapses to `max(a, b)`.

With `_sfpu_exp_fp32_accurate_` instead, same ten cases, worst relative error **3.09e-08** — the
`a = b = 0` case comes out at `0.693147`.

Two smaller things while I was in there:

- `_calculate_exponential_body_<is_fp32_dest_acc_en>` passes the destination-accumulation flag
  into the `APPROXIMATION_MODE` template slot. They are different things, and the effect is
  inverted from what you would want: an fp32 destination selects the *approximate* body.
- `_sfpu_exp_fp32_accurate_unsafe_` looks like the cheaper option but is not usable here — it
  drops the underflow guard, and `-|a-b|` reaches large negative values. It returns `177.4` for
  `logaddexp(100, 0)`.

On method, so the numbers are checkable: I could not drive `logaddexp` directly without building
the host side, so I compiled your kernel body into the `XLOGY` branch of `calculate_sfpu_binary` —
the same function and the same position as your change, and the branch that already uses the
`_calculate_log_body_(0, dst_index_out)` pattern — and called `ttnn.xlogy`. My first attempt put it
in `calculate_sfpu_binary_div` instead and produced nonsense, so the placement matters; the table
above is from the faithful one.

I have a branch with the accurate-exp variant and a ten-case test if it saves you time, but it is
your PR and a one-line swap is probably simpler than merging anything.

**Comment by thwaitsusa0316** [comment]
Eazyhood....  follow up:  took the swap. `09941dc` moves both ops to
`_sfpu_exp_fp32_accurate_`
on both arches, template argument omitted so it keeps the underflow guard.
You were right about`_calculate_exponential_body_<is_fp32_dest_acc_en>` landing
in the `APPROXIMATION_MODE` slot too — real bug, independent of
accuracy. Credited you in the commit. My earlier reply went out by email and was terser than your
review deserved.  Apologies steve

File: tt_metal/tt-llk/tt_llk_wormhole_b0/common/inc/ckernel_defs.h [hunk 1]
```diff
@@ -308,6 +308,8 @@ enum class BinaryOp : std::uint8_t
     MAX             = 17,
     MIN             = 18,
     FMOD            = 19,
+    LOGADDEXP       = 20,
+    LOGADDEXP2      = 21,
     REMAINDER       = 20,
     BITWISE_AND     = 21,
```
Comment (ldjurovicTT) [inline tt_metal/tt-llk/tt_llk_wormhole_b0/common/inc/ckernel_defs.h]: Replicating numbers

File: tt_metal/tt-llk/tt_llk_blackhole/common/inc/ckernel_defs.h [hunk 2]
```diff
@@ -326,6 +326,8 @@ enum class BinaryOp : std::uint8_t
     MAX             = 17,
     MIN             = 18,
     FMOD            = 19,
+    LOGADDEXP       = 20,
+    LOGADDEXP2      = 21,
     REMAINDER       = 20,
     BITWISE_AND     = 21,
```
Comment (ldjurovicTT) [inline tt_metal/tt-llk/tt_llk_blackhole/common/inc/ckernel_defs.h]: Replicating numbers

File: tt_metal/tt-llk/tt_llk_wormhole_b0/common/inc/ckernel_defs.h (hunk 1 above)
Comment (thwaitsusa0316) [reply tt_metal/tt-llk/tt_llk_wormhole_b0/common/inc/ckernel_defs.h]: Good catch — 20 and 21 were already taken by `REMAINDER` and `BITWISE_AND`, so
`BinaryOp::LOGADDEXP == BinaryOp::REMAINDER` as written. It compiled because
duplicate enumerator values are legal, but any dispatch testing one would have
matched the other.

Fixed in 3fb7f67: moved to 43/44, appended after `FMOD_INT32 = 42` per the
"appended at the end so existing values are unchanged" convention that block
already documents. No existing enumerator moves, both architectures.
